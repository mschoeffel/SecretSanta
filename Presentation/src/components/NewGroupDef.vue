<template>
  <v-container class="fill-height" fluid>
    <v-row align="center" justify="center">
      <v-col cols="12" sm="8" md="4">
        <v-card class="elevation-12">
          <v-toolbar color="primary" dark flat>
            <v-toolbar-title>{{ $t('new-group.formtitle')}}</v-toolbar-title>
          </v-toolbar>
          <v-card-text>
            <v-stepper v-model="step" alt-labels>
              <v-stepper-header>
                <v-stepper-step key="group" step="1">{{ $t('new-group.group')}}</v-stepper-step>
                <v-divider></v-divider>
                <v-stepper-step key="member" step="2">{{ $t('new-group.member')}}</v-stepper-step>
              </v-stepper-header>
              <v-stepper-items>
                <v-stepper-content key="group" step="1">
                  <v-form>
                      <v-text-field
                        :label="$t('new-group.name')"
                        :error-messages="$t(groupnameerror)"
                        name="name"
                        v-model="groupname"
                        v-on:change="checkGroupName"
                        prepend-icon="mdi-account-group"
                        type="text"
                      />
                    <v-text-field
                      :label="$t('new-group.membercount')"
                      v-model.number="membercount"
                      name="name"
                      prepend-icon="mdi-numeric"
                      type="number"
                    />
                    <v-divider></v-divider>
                    <v-checkbox
                      :disabled="!allowEmail"
                      v-model="checkbox"
                      name="email"
                      color="primary"
                      :label="$t('new-group.email')"
                      prepend-icon="mdi-email-lock"
                    ></v-checkbox>
                    <v-divider></v-divider>
                    <v-checkbox
                      v-model="checkboxRerolls"
                      color="primary"
                      :label="$t('new-group.rerolls')"
                      prepend-icon="mdi-sync"
                    ></v-checkbox>
                    <v-text-field
                      :label="$t('new-group.rerolls-slider')"
                      name="rerolls"
                      v-model="rerolls"
                      prepend-icon="mdi-sync"
                      :disabled="!checkboxRerolls"
                      type="number"
                    />
                  </v-form>
                </v-stepper-content>
                <v-stepper-content key="member" step="2">
                  <v-list>
                    <div v-for="member, index in members">
                      <v-subheader>{{index + 1}}</v-subheader>
                      <v-list-item inactive="true" selectable="false">
                        <v-list-item-content>
                          <v-text-field
                            v-model="member.name"
                            :label="$t('new-group.member-name')"
                            name="name"
                            prepend-icon="mdi-account"
                            type="text"
                          />
                        </v-list-item-content>
                      </v-list-item>
                      <v-list-item v-if="checkbox" inactive="true" selectable="false">
                        <v-list-item-content>
                          <v-text-field
                            v-model="member.email"
                            :label="$t('new-group.member-email')"
                            name="name"
                            prepend-icon="mdi-email"
                            type="text"
                          />
                        </v-list-item-content>
                      </v-list-item>
                      <v-divider v-if="n < membercount"></v-divider>
                    </div>
                  </v-list>
                </v-stepper-content>
              </v-stepper-items>
            </v-stepper>
          </v-card-text>
          <v-card-actions>
            <v-btn color="secondary" v-on:click="stepBack">{{ $tc('new-group.back', step)}}</v-btn>
            <v-spacer />
            <v-btn color="primary" v-on:click="stepForward">{{ $tc('new-group.next', step)}}</v-btn>
          </v-card-actions>
        </v-card>
      </v-col>
    </v-row>
  </v-container>
</template>

<script>
import api from "./backend-api";

export default {
  props: {
    source: String
  },
  data: () => ({
    groupname: "",
    groupnameerror: "",
    rerolls: 0,
    showPassword: false,
    allowEmail: false,
    min: 0,
    max: 50,
    slider: 10,
    checkbox: false,
    checkboxRerolls: false,
    sliderRerolls: 0,
    minRerolls: 0,
    maxRerolls: 5,
    step: 1,
    membercount: 0,
    membercountsave: 0,
    members: []
  }),
  methods: {
    stepBack: function() {
      if (this.step === 1) {
        this.$router.push({ path: "/" });
      } else {
        this.step -= 1;
      }
    },
    checkGroupName: function(){
        api.checkGroupName(this.groupname)
        .then(response => {
            console.log(response);
            if(response.data == true){
                this.groupnameerror = "new-group.error-groupname-exists";
            } else{
                this.groupnameerror = "";
            }
        })
    },
    stepForward: function() {
      if (this.step === 2) {
        api.createGroup(this.groupname, this.rerolls, this.members)
        .then(response => {
             console.log(response);
             this.$router.push({ path: "/group" });
        })
        .catch(
            error => {
            console.log(error.response);
            if(error.response.status == 409){
                this.groupnameerror = "new-group.error-groupname-exists";
                this.step = 1;
            }
        });

      } else {
        if (this.membercountsave >= this.membercount) {
          this.members.splice(this.membercount);
          this.membercountsave = this.membercount;
        } else if (this.membercountsave <= this.membercount) {
          let memberstemp = Array.from(
            { length: this.membercount - this.membercountsave },
            () => ({ name: "", email: "" })
          );
          this.members = this.members.concat(memberstemp);
          this.membercountsave = this.membercount;
        }
        this.step += 1;
      }
    }
  }
};
</script>