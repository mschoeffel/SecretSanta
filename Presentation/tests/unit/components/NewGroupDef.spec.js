import Vue from 'vue'
import Vuetify from 'vuetify'
import VueI18n from 'vue-i18n'

import NewGroupDef from '@/components/NewGroupDef'

import { mount, createLocalVue } from '@vue/test-utils';

const localVue = createLocalVue()

Vue.use(VueI18n)
const i18n = new VueI18n({});

describe('NewGroupDef.vue', () => {
  let vuetify

  beforeEach(() => {
    vuetify = new Vuetify()
  })

  const mountFunction = options => {
    return mount(NewGroupDef, {
      localVue,
      vuetify,
      i18n,
      ...options,
    })
  }

  it('should render correct group creation form', () => {
    const wrapper = mountFunction()

    expect(wrapper.html()).toMatchSnapshot()
    expect(wrapper.vm.step).toBe(1)
  })

  it('should give invalid input to empty form', () =>{
    const wrapper = mountFunction()

    wrapper.find('v-btn[name=next]').trigger('click')

    expect(wrapper.vm.step).toBe(1)
    expect(wrapper.vm.groupnameerror).toBe("new-group.validation-groupnamerequired")
    expect(wrapper.vm.validationerror).toBe("new-group.validationerror")
  })
  
  
  it('should give invalid input to empty groupname', () =>{
    const wrapper = mountFunction()

    wrapper.setData({groupname: ""})
    wrapper.find('v-btn[name=next]').trigger('click')

    expect(wrapper.vm.step).toBe(1)
    expect(wrapper.vm.groupnameerror).toBe("new-group.validation-groupnamerequired")
    expect(wrapper.vm.validationerror).toBe("new-group.validationerror")
  })

  
  it('should give invalid input to too long groupname', () =>{
    const wrapper = mountFunction()

    wrapper.setData({groupname: "0123456789123456"})
    wrapper.find('v-btn[name=next]').trigger('click')

    expect(wrapper.vm.step).toBe(1)
    expect(wrapper.vm.groupnameerror).toBe("new-group.validation-groupnamelength")
    expect(wrapper.vm.validationerror).toBe("new-group.validationerror")
  })

  
  it('should give invalid input to 0 member', () =>{
    const wrapper = mountFunction()

    wrapper.setData({membercount: 0})
    wrapper.find('v-btn[name=next]').trigger('click')

    expect(wrapper.vm.step).toBe(1)
    expect(wrapper.vm.membercounterror).toBe("new-group.validation-membercountbounds")
    expect(wrapper.vm.validationerror).toBe("new-group.validationerror")
  })

  it('should give invalid input to 1 member', () =>{
    const wrapper = mountFunction()

    wrapper.setData({membercount: 1})
    wrapper.find('v-btn[name=next]').trigger('click')

    expect(wrapper.vm.step).toBe(1)
    expect(wrapper.vm.membercounterror).toBe("new-group.validation-membercountbounds")
    expect(wrapper.vm.validationerror).toBe("new-group.validationerror")
  })

  it('should give invalid input to too much member', () =>{
    const wrapper = mountFunction()

    wrapper.setData({membercount: 51})
    wrapper.find('v-btn[name=next]').trigger('click')

    expect(wrapper.vm.step).toBe(1)
    expect(wrapper.vm.membercounterror).toBe("new-group.validation-membercountbounds")
    expect(wrapper.vm.validationerror).toBe("new-group.validationerror")
  })

  it('should give invalid input to negative rerolls', () =>{
    const wrapper = mountFunction()

    wrapper.setData({rerolls: -1})
    wrapper.find('v-btn[name=next]').trigger('click')

    expect(wrapper.vm.step).toBe(1)
    expect(wrapper.vm.rerollserror).toBe("new-group.validation-rerollsbounds")
    expect(wrapper.vm.validationerror).toBe("new-group.validationerror")
  })

  it('should give invalid input to too much rerolls', () =>{
    const wrapper = mountFunction()

    wrapper.setData({rerolls: 101})
    wrapper.find('v-btn[name=next]').trigger('click')

    expect(wrapper.vm.step).toBe(1)
    expect(wrapper.vm.rerollserror).toBe("new-group.validation-rerollsbounds")
    expect(wrapper.vm.validationerror).toBe("new-group.validationerror")
  })

})
