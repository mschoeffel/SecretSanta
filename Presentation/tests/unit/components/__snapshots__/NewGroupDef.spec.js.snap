// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewGroupDef.vue should render correct group creation form 1`] = `
<v-container fluid="" class="fill-height">
  <v-row align="center" justify="center">
    <v-col cols="12" sm="12" md="8">
      <v-card class="elevation-12">
        <v-toolbar color="primary" dark="" flat="">
          <v-toolbar-title>new-group.formtitle</v-toolbar-title>
          <!---->
          <v-spacer></v-spacer>
          <v-btn icon="">
            <v-icon>mdi-help</v-icon>
          </v-btn>
        </v-toolbar>
        <v-card-text>
          <v-stepper alt-labels="">
            <v-stepper-header>
              <v-stepper-step step="1">new-group.group</v-stepper-step>
              <v-divider></v-divider>
              <v-stepper-step step="2">new-group.member</v-stepper-step>
              <v-divider></v-divider>
              <v-stepper-step step="3">new-group.result</v-stepper-step>
            </v-stepper-header>
            <v-stepper-items>
              <v-stepper-content step="1">
                <v-alert name="info" id="info" type="info" outlined="" dismissible="">new-group.info1</v-alert>
                <!---->
                <v-text-field label="new-group.name" name="groupname" id="groupname" type="text" prepend-icon="mdi-account-group" counter="15" error-messages=""></v-text-field>
                <v-text-field label="new-group.membercount" name="name" id="name" type="number" prepend-icon="mdi-numeric" min="2" max="50" error-messages=""></v-text-field>
                <v-divider></v-divider>
                <v-checkbox label="new-group.rerolls" name="enableRerolls" id="enableRerolls" prepend-icon="mdi-sync" color="primary"></v-checkbox>
                <v-text-field label="new-group.rerolls-slider" name="rerolls" id="rerolls" type="number" prepend-icon="mdi-sync" min="0" max="100" error-messages="" disabled="true"></v-text-field>
              </v-stepper-content>
              <v-stepper-content step="2">
                <v-alert name="info" id="info" type="info" outlined="" dismissible="">new-group.info2</v-alert>
                <!---->
                <v-simple-table fixed-header="">
                  <thead>
                    <tr>
                      <th class="text-left">new-group.member-number</th>
                      <th class="text-left">new-group.member-name</th>
                    </tr>
                  </thead>
                  <tbody></tbody>
                </v-simple-table>
              </v-stepper-content>
              <v-stepper-content step="3">
                <v-alert name="info" id="info" type="info" outlined="" dismissible="">new-group.info3</v-alert>
                <!---->
              </v-stepper-content>
            </v-stepper-items>
          </v-stepper>
        </v-card-text>
        <v-card-actions>
          <v-btn color="secondary" name="back" id="back">new-group.back</v-btn>
          <v-spacer></v-spacer>
          <v-btn color="primary" name="next" id="next">new-group.next</v-btn>
        </v-card-actions>
      </v-card>
    </v-col>
  </v-row>
</v-container>
`;
